### PR TITLE
NN-1202: Add in expired password functionality

### PR DESCRIPTION
**Pull request by petergphillips**
First stab at expired password functionality.  Will also work for users that don't have expired passwords.

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -108,10 +108,9 @@ dependencies {
     testCompile('org.springframework.security:spring-security-test')
     testRuntime 'com.h2database:h2:1.4.196'
 
-    testCompile 'org.assertj:assertj-core:3.8.0'
```
Comment (petergphillips) [inline build.gradle]: don't need to pin assertj as already included

File: src/main/java/uk/gov/justice/digital/hmpps/oauth2server/security/JwtAuthenticationHelper.java [hunk 2]
```diff
@@ -62,9 +63,10 @@ String createJwt(final Authentication authentication) {
                     .getBody();
             final String username = body.getSubject();
             final String authoritiesString = body.get("authorities", String.class);
-            final Collection<? extends GrantedAuthority> authorities = Stream.of(authoritiesString.split(",")).
-                    map(SimpleGrantedAuthority::new).
-                    collect(Collectors.toList());
+            final Collection<? extends GrantedAuthority> authorities = Stream.of(authoritiesString.split(","))
+                    .filter(StringUtils::isNotEmpty)
```
Comment (petergphillips) [inline src/main/java/uk/gov/justice/digital/hmpps/oauth2server/security/JwtAuthenticationHelper.java]: need to cope with users without any authorities e.g ZQU69A

File: src/main/java/uk/gov/justice/digital/hmpps/oauth2server/security/OracleUserService.java [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    private final JdbcTemplate jdbcTemplate;
+
+    public OracleUserService(@Qualifier("dataSource") final DataSource dataSource) {
+        jdbcTemplate = new JdbcTemplate(dataSource);
+    }
+
+    @Transactional
+    public void changePassword(final String username, final String password) {
+        try {
+            jdbcTemplate.update(String.format("ALTER USER %s IDENTIFIED BY %s", username, password));
```
Comment (petergphillips) [inline src/main/java/uk/gov/justice/digital/hmpps/oauth2server/security/OracleUserService.java]: note couldn't use parameters here

File: src/main/java/uk/gov/justice/digital/hmpps/oauth2server/security/UserDetailsServiceImpl.java [hunk 4]
```diff
@@ -39,25 +37,25 @@ public UserDetailsServiceImpl(UserService userService,
...
+        final var userByUsername = userService.getUserByUsername(username).orElseThrow(() -> new UsernameNotFoundException(username));
         if (userByUsername.filterByCaseload(apiCaseloadId).isEmpty()) {
             throw new UnapprovedClientAuthenticationException(format("User does not have access to caseload %s", apiCaseloadId));
         }
 
-        Set<GrantedAuthority> authorities = userByUsername.filterRolesByCaseload(apiCaseloadId).stream()
+        final Set<GrantedAuthority> authorities = userByUsername.filterRolesByCaseload(apiCaseloadId).stream()
                 .filter(Objects::nonNull)
-                .map(role -> new SimpleGrantedAuthority("ROLE_" + StringUtils.upperCase(StringUtils.replaceAll(role.getRole().getCode(), "-", "_"))))
+                .map(role -> new SimpleGrantedAuthority("ROLE_" + StringUtils.upperCase(RegExUtils.replaceAll(role.getRole().getCode(), "-", "_"))))
```
Comment (petergphillips) [inline src/main/java/uk/gov/justice/digital/hmpps/oauth2server/security/UserDetailsServiceImpl.java]: string utils version was deprecated, so use replacement

File: src/main/java/uk/gov/justice/digital/hmpps/oauth2server/security/OracleAuthenticationProvider.java [hunk 5]
```diff
@@ -75,38 +74,38 @@ protected void additionalAuthenticationChecks(final UserDetails userDetails, fin
...
+            throw new BadCredentialsException("Authentication failed: password does not match stored value");
         }
     }
 
     private String encode(final String rawPassword, final String salt) {
-        final List<SqlParameter> params = new ArrayList<>();
-        params.add(new SqlOutParameter("encodedPassword", Types.VARCHAR));
-        params.add(new SqlParameter(Types.VARCHAR, salt));
-        params.add(new SqlParameter(Types.VARCHAR, rawPassword));
+        final List<SqlParameter> params = List.of(
```
Comment (Mjwillis) [inline src/main/java/uk/gov/justice/digital/hmpps/oauth2server/security/OracleAuthenticationProvider.java]: much nicer!

**Pull request merged**